File: ts/test/entity/universal/UniversalDirect.test.ts

const envlocal = __dirname + '/../../../.env.local'
require('dotenv').config({ quiet: true, path: [envlocal] })

import { test, describe } from 'node:test'
import assert from 'node:assert'


import { UniversalManager, UniversalSDK, stdutil } from '../../..'

import {
  envOverride,
} from '../../utility'


describe('UniversalDirect', async () => {

  const um = new UniversalManager({ registry: __dirname + '/../../../test/registry' })
  const sdk = um.make('voxgig-solardemo')
  const entityMap: any = sdk._config.entity
  const struct = sdk.utility().struct
  const items = struct.items

  const live = 'TRUE' === process.env.UNIVERSAL_TEST_LIVE


  test('direct-exists', async () => {
    const setup = directSetup(um, sdk)
    assert('function' === typeof setup.client.direct)
    assert('function' === typeof setup.client.prepare)
  })


  items(entityMap, (item: any[]) => {
    const entityDef = item[1]
    const entityName = entityDef.name
    const ops = entityDef.op || {}

    const hasLoad = null != ops.load
    const hasList = null != ops.list

    if (!hasLoad && !hasList) {
      return
    }

    if (hasLoad) {
      const loadOp = ops.load
      const loadTarget = loadOp.targets?.[0]

      if (null != loadTarget) {
        test('direct-load-' + entityName, async () => {
          const loadPath = (loadTarget.parts || []).join('/')
          const loadParams = loadTarget.args?.params || []

          if (live) {
            const idmap = await resolveIdmap(um, sdk, entityName, entityMap)
            const setup = directSetup(um, sdk)

            // First list to discover a real entity ID.
            if (hasList) {
              const listTarget = ops.list.targets?.[0]
              if (null != listTarget) {
                const listPath = (listTarget.parts || []).join('/')
                const listParams = listTarget.args?.params || []

                // Try multiple parent refs to find one with child entities.
                let found: any = null
                let lparams: any = {}
                for (let t = 0; t < 3 && null == found; t++) {
                  lparams = {}
                  for (const p of listParams) {
                    const ref = p.name.replace(/_id$/, '') +
                      String(t).padStart(2, '0')
                    lparams[p.name] = idmap[ref] || ref
                  }

                  const listResult: any = await setup.client.direct({
                    path: listPath,
                    method: 'GET',
                    params: lparams,
                  })

                  assert(listResult.ok === true)
                  assert(Array.isArray(listResult.data))

                  if (listResult.data.length >= 1) {
                    found = listResult.data[0]
                  }
                }

                if (null != found) {
                  const params: any = {}
                  for (const p of loadParams) {
                    params[p.name] = found[p.name] || lparams[p.name]
                  }

                  const result: any = await setup.client.direct({
                    path: loadPath,
                    method: 'GET',
                    params,
                  })

                  assert(result.ok === true)
                  assert(result.status === 200)
                  assert(null != result.data)
                  assert(result.data.id === found.id)
                }
              }
            }
          }
          else {
            const setup = directSetup(um, sdk, { id: 'direct01' })
            const { client, calls } = setup

            const params: any = {}
            for (let i = 0; i < loadParams.length; i++) {
              params[loadParams[i].name] = 'direct0' + (i + 1)
            }

            const result: any = await client.direct({
              path: loadPath,
              method: 'GET',
              params,
            })

            assert(result.ok === true)
            assert(result.status === 200)
            assert(null != result.data)
            assert(result.data.id === 'direct01')

            assert(calls.length === 1)
            assert(calls[0].init.method === 'GET')

            for (let i = 0; i < loadParams.length; i++) {
              assert(calls[0].url.includes('direct0' + (i + 1)))
            }
          }
        })
      }
    }

    if (hasList) {
      const listOp = ops.list
      const listTarget = listOp.targets?.[0]

      if (null != listTarget) {
        test('direct-list-' + entityName, async () => {
          const listPath = (listTarget.parts || []).join('/')
          const listParams = listTarget.args?.params || []

          if (live) {
            const idmap = await resolveIdmap(um, sdk, entityName, entityMap)
            const setup = directSetup(um, sdk)

            // For entities with parent params, try each known parent
            // to find one that has child entities.
            let found = false
            const maxTries = listParams.length > 0 ? 3 : 1
            for (let t = 0; t < maxTries && !found; t++) {
              const params: any = {}
              for (const p of listParams) {
                const base = (p.name === 'id' ? entityName : p.name.replace(/_id$/, ''))
                const ref = base + String(t).padStart(2, '0')
                params[p.name] = idmap[ref] || ref
              }

              const result: any = await setup.client.direct({
                path: listPath,
                method: 'GET',
                params,
              })

              assert(result.ok === true)
              assert(result.status === 200)
              assert(Array.isArray(result.data))

              if (result.data.length >= 1) {
                found = true
              }
            }

            if (listParams.length === 0) {
              assert(found, 'expected at least one entity in list')
            }
          }
          else {
            const setup = directSetup(um, sdk, [{ id: 'direct01' }, { id: 'direct02' }])
            const { client, calls } = setup

            const params: any = {}
            for (let i = 0; i < listParams.length; i++) {
              params[listParams[i].name] = 'direct0' + (i + 1)
            }

            const result: any = await client.direct({
              path: listPath,
              method: 'GET',
              params,
            })

            assert(result.ok === true)
            assert(result.status === 200)
            assert(Array.isArray(result.data))
            assert(result.data.length === 2)

            assert(calls.length === 1)
            assert(calls[0].init.method === 'GET')

            for (let i = 0; i < listParams.length; i++) {
              assert(calls[0].url.includes('direct0' + (i + 1)))
            }
          }
        })
      }
    }
  })

})



async function resolveIdmap(um: any, sdk: any, entityName: string, entityMap: any): Promise<any> {
  const clientStruct = sdk.utility().struct
  const items = clientStruct.items
  const transform = clientStruct.transform

  const idEntries: string[] = []
  items(entityMap, (item: any[]) => {
    const ename = item[1].name
    for (let i = 0; i < 3; i++) {
      idEntries.push(`${ename}${String(i).padStart(2, '0')}`)
    }
  })

  let idmap = transform(
    idEntries,
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'UNIVERSAL_TEST_ENTID': idmap,
    'UNIVERSAL_TEST_LIVE': 'FALSE',
  })

  idmap = env['UNIVERSAL_TEST_ENTID']

  // In live mode, discover real parent entity IDs by listing parent entities.
  if ('TRUE' === process.env.UNIVERSAL_TEST_LIVE) {
    const liveClient = new UniversalSDK(um, {
      ref: 'voxgig-solardemo',
      model: sdk._options.model,
    })

    const discoveries: Promise<void>[] = []
    items(entityMap, (item: any[]) => {
      const eDef = item[1]
      const eName = eDef.name
      const listOp = eDef.op?.list
      const listTarget = listOp?.targets?.[0]
      if (null == listTarget) return

      const listParams = listTarget.args?.params || []
      if (listParams.length > 0) return // skip nested entities in discovery

      const listPath = (listTarget.parts || []).join('/')
      discoveries.push((async () => {
        const res: any = await liveClient.direct({ path: listPath, method: 'GET', params: {} })
        if (res.ok && Array.isArray(res.data)) {
          for (let i = 0; i < Math.min(res.data.length, 3); i++) {
            const ref = `${eName}${String(i).padStart(2, '0')}`
            idmap[ref] = res.data[i].id
          }
        }
      })())
    })

    await Promise.all(discoveries)
  }

  return idmap
}


function directSetup(um: any, sdk: any, mockres?: any) {
  const live = 'TRUE' === process.env.UNIVERSAL_TEST_LIVE

  if (live) {
    const client = new UniversalSDK(um, {
      ref: 'voxgig-solardemo',
      model: sdk._options.model,
    })
    return { client, calls: [] as any[], live: true }
  }

  const calls: any[] = []

  const mockFetch = async (url: string, init: any) => {
    calls.push({ url, init })
    return {
      status: 200,
      statusText: 'OK',
      headers: {},
      json: async () => (null != mockres ? mockres : { id: 'direct01' }),
    }
  }

  const client = new UniversalSDK(um, {
    ref: 'voxgig-solardemo',
    model: sdk._options.model,
    base: 'http://localhost:8080',
    system: { fetch: mockFetch },
  })

  return { client, calls, live: false }
}

File: go/test/entity/universal/universal_direct_test.go
package universal

import (
	"fmt"
	"os"
	"strings"
	"testing"

	sdk "voxgiguniversalsdk"
	"voxgiguniversalsdk/core"

	vs "github.com/voxgig/struct"
)

func TestUniversalDirect(t *testing.T) {
	um := sdk.NewUniversalManager(map[string]any{
		"registry": "../../../test/registry",
	})
	baseSDK := um.Make("voxgig-solardemo")
	entityMap := core.ToMapAny(vs.GetPath([]any{"entity"}, baseSDK.GetRootCtx().Config))

	live := os.Getenv("UNIVERSAL_TEST_LIVE") == "TRUE"

	t.Run("direct-exists", func(t *testing.T) {
		setup := directSetup(um, baseSDK, nil)
		if setup.client == nil {
			t.Fatal("client should not be nil")
		}
	})

	for _, item := range vs.Items(entityMap) {
		entDef := core.ToMapAny(item[1])
		if entDef == nil {
			continue
		}
		entityName, _ := vs.GetProp(entDef, "name").(string)
		ops := core.ToMapAny(vs.GetProp(entDef, "op"))
		if ops == nil {
			continue
		}

		hasLoad := vs.GetProp(ops, "load") != nil
		hasList := vs.GetProp(ops, "list") != nil

		if !hasLoad && !hasList {
			continue
		}

		if hasLoad {
			loadOp := core.ToMapAny(vs.GetProp(ops, "load"))
			targets, _ := vs.GetProp(loadOp, "targets").([]any)
			if len(targets) == 0 {
				continue
			}
			loadTarget := core.ToMapAny(targets[0])
			if loadTarget == nil {
				continue
			}

			name := entityName
			t.Run("direct-load-"+name, func(t *testing.T) {
				parts, _ := vs.GetProp(loadTarget, "parts").([]any)
				loadPath := vs.Join(parts, "/", true)
				loadParams, _ := vs.GetPath([]any{"args", "params"}, loadTarget).([]any)

				if live {
					idmap := resolveIdmap(um, baseSDK, entityMap)
					setup := directSetup(um, baseSDK, nil)

					if hasList {
						listOp := core.ToMapAny(vs.GetProp(ops, "list"))
						listTargets, _ := vs.GetProp(listOp, "targets").([]any)
						if len(listTargets) > 0 {
							listTarget := core.ToMapAny(listTargets[0])
							if listTarget != nil {
								listParts, _ := vs.GetProp(listTarget, "parts").([]any)
								listPath := vs.Join(listParts, "/", true)
								listParams, _ := vs.GetPath([]any{"args", "params"}, listTarget).([]any)

								// Try multiple parent refs to find one with child entities.
								var found map[string]any
								lparams := map[string]any{}
								for tr := 0; tr < 3 && found == nil; tr++ {
									lparams = map[string]any{}
									for _, p := range listParams {
										pm := core.ToMapAny(p)
										if pm == nil {
											continue
										}
										pname, _ := vs.GetProp(pm, "name").(string)
										ref := strings.TrimSuffix(pname, "_id") +
											fmt.Sprintf("%02d", tr)
										if idmap[ref] != nil {
											lparams[pname] = idmap[ref]
										} else {
											lparams[pname] = ref
										}
									}

									listResult, err := setup.client.Direct(map[string]any{
										"path": listPath, "method": "GET", "params": lparams,
									})
									if err != nil {
										t.Fatalf("direct list failed: %v", err)
									}
									if listResult["ok"] != true {
										t.Fatalf("expected ok=true, got %v", listResult["ok"])
									}
									if dataList, ok := listResult["data"].([]any); ok && len(dataList) >= 1 {
										found = core.ToMapAny(dataList[0])
									}
								}

								if found != nil {
									params := map[string]any{}
									for _, p := range loadParams {
										pm := core.ToMapAny(p)
										if pm == nil {
											continue
										}
										pname, _ := vs.GetProp(pm, "name").(string)
										if found[pname] != nil {
											params[pname] = found[pname]
										} else if lparams[pname] != nil {
											params[pname] = lparams[pname]
										}
									}

									result, err := setup.client.Direct(map[string]any{
										"path": loadPath, "method": "GET", "params": params,
									})
									if err != nil {
										t.Fatalf("direct load failed: %v", err)
									}
									if result["ok"] != true {
										t.Fatalf("expected ok=true, got %v", result["ok"])
									}
									if core.ToInt(result["status"]) != 200 {
										t.Fatalf("expected status 200, got %v", result["status"])
									}
									if result["data"] == nil {
										t.Fatal("expected data to be non-nil")
									}
									if dm := core.ToMapAny(result["data"]); dm != nil {
										if dm["id"] != found["id"] {
											t.Fatalf("expected id=%v, got %v", found["id"], dm["id"])
										}
									}
								}
							}
						}
					}
				} else {
					setup := directSetup(um, baseSDK, map[string]any{"id": "direct01"})

					params := map[string]any{}
					for i, p := range loadParams {
						pm := core.ToMapAny(p)
						if pm != nil {
							pname, _ := vs.GetProp(pm, "name").(string)
							params[pname] = fmt.Sprintf("direct0%d", i+1)
						}
					}

					result, err := setup.client.Direct(map[string]any{
						"path": loadPath, "method": "GET", "params": params,
					})
					if err != nil {
						t.Fatalf("direct failed: %v", err)
					}
					if result["ok"] != true {
						t.Fatalf("expected ok=true, got %v", result["ok"])
					}
					if core.ToInt(result["status"]) != 200 {
						t.Fatalf("expected status 200, got %v", result["status"])
					}
					if result["data"] == nil {
						t.Fatal("expected data to be non-nil")
					}
					if dm := core.ToMapAny(result["data"]); dm != nil {
						if dm["id"] != "direct01" {
							t.Fatalf("expected data.id=direct01, got %v", dm["id"])
						}
					}

					if len(*setup.calls) != 1 {
						t.Fatalf("expected 1 call, got %d", len(*setup.calls))
					}
					call := (*setup.calls)[0]
					if initMap, ok := call["init"].(map[string]any); ok {
						if initMap["method"] != "GET" {
							t.Fatalf("expected method GET, got %v", initMap["method"])
						}
					}
					for i := range loadParams {
						url, _ := call["url"].(string)
						if !strings.Contains(url, fmt.Sprintf("direct0%d", i+1)) {
							t.Fatalf("expected url to contain direct0%d, got %v", i+1, url)
						}
					}
				}
			})
		}

		if hasList {
			listOp := core.ToMapAny(vs.GetProp(ops, "list"))
			targets, _ := vs.GetProp(listOp, "targets").([]any)
			if len(targets) == 0 {
				continue
			}
			listTarget := core.ToMapAny(targets[0])
			if listTarget == nil {
				continue
			}

			name := entityName
			t.Run("direct-list-"+name, func(t *testing.T) {
				parts, _ := vs.GetProp(listTarget, "parts").([]any)
				listPath := vs.Join(parts, "/", true)
				listParams, _ := vs.GetPath([]any{"args", "params"}, listTarget).([]any)

				if live {
					idmap := resolveIdmap(um, baseSDK, entityMap)
					setup := directSetup(um, baseSDK, nil)

					// For entities with parent params, try each known parent
					// to find one that has child entities.
					found := false
					maxTries := 1
					if len(listParams) > 0 {
						maxTries = 3
					}
					for tr := 0; tr < maxTries && !found; tr++ {
						params := map[string]any{}
						for _, p := range listParams {
							pm := core.ToMapAny(p)
							if pm == nil {
								continue
							}
							pname, _ := vs.GetProp(pm, "name").(string)
							base := pname
							if pname == "id" {
								base = entityName
							} else {
								base = strings.TrimSuffix(pname, "_id")
							}
							ref := fmt.Sprintf("%s%02d", base, tr)
							if idmap[ref] != nil {
								params[pname] = idmap[ref]
							} else {
								params[pname] = ref
							}
						}

						result, err := setup.client.Direct(map[string]any{
							"path": listPath, "method": "GET", "params": params,
						})
						if err != nil {
							t.Fatalf("direct failed: %v", err)
						}
						if result["ok"] != true {
							t.Fatalf("expected ok=true, got %v", result["ok"])
						}
						if core.ToInt(result["status"]) != 200 {
							t.Fatalf("expected status 200, got %v", result["status"])
						}
						if dataList, ok := result["data"].([]any); ok && len(dataList) >= 1 {
							found = true
						}
					}

					if len(listParams) == 0 && !found {
						t.Fatal("expected at least one entity in list")
					}
				} else {
					mockData := []any{
						map[string]any{"id": "direct01"},
						map[string]any{"id": "direct02"},
					}
					setup := directSetup(um, baseSDK, mockData)

					params := map[string]any{}
					for i, p := range listParams {
						pm := core.ToMapAny(p)
						if pm != nil {
							pname, _ := vs.GetProp(pm, "name").(string)
							params[pname] = fmt.Sprintf("direct0%d", i+1)
						}
					}

					result, err := setup.client.Direct(map[string]any{
						"path": listPath, "method": "GET", "params": params,
					})
					if err != nil {
						t.Fatalf("direct failed: %v", err)
					}
					if result["ok"] != true {
						t.Fatalf("expected ok=true, got %v", result["ok"])
					}
					if core.ToInt(result["status"]) != 200 {
						t.Fatalf("expected status 200, got %v", result["status"])
					}
					if dataList, ok := result["data"].([]any); ok {
						if len(dataList) != 2 {
							t.Fatalf("expected 2 items, got %d", len(dataList))
						}
					} else {
						t.Fatalf("expected data to be an array, got %T", result["data"])
					}

					if len(*setup.calls) != 1 {
						t.Fatalf("expected 1 call, got %d", len(*setup.calls))
					}
					call := (*setup.calls)[0]
					if initMap, ok := call["init"].(map[string]any); ok {
						if initMap["method"] != "GET" {
							t.Fatalf("expected method GET, got %v", initMap["method"])
						}
					}

					for i := range listParams {
						url, _ := call["url"].(string)
						if !strings.Contains(url, fmt.Sprintf("direct0%d", i+1)) {
							t.Fatalf("expected url to contain direct0%d, got %v", i+1, url)
						}
					}
				}
			})
		}
	}
}

type directSetupResult struct {
	client *sdk.UniversalSDK
	calls  *[]map[string]any
	live   bool
}

func directSetup(um *sdk.UniversalManager, baseSDK *sdk.UniversalSDK, mockres any) *directSetupResult {
	live := os.Getenv("UNIVERSAL_TEST_LIVE") == "TRUE"

	if live {
		client := sdk.NewUniversalSDK(um, map[string]any{
			"ref":   "voxgig-solardemo",
			"model": um.ResolveModel("voxgig-solardemo"),
		})
		return &directSetupResult{client: client, calls: &[]map[string]any{}, live: true}
	}

	calls := &[]map[string]any{}

	mockFetch := func(url string, init map[string]any) (map[string]any, error) {
		*calls = append(*calls, map[string]any{"url": url, "init": init})
		return map[string]any{
			"status":     200,
			"statusText": "OK",
			"headers":    map[string]any{},
			"json": (func() any)(func() any {
				if mockres != nil {
					return mockres
				}
				return map[string]any{"id": "direct01"}
			}),
		}, nil
	}

	client := core.NewUniversalSDK(um, map[string]any{
		"model": um.ResolveModel("voxgig-solardemo"),
		"ref":   "voxgig-solardemo",
		"base":  "http://localhost:8080",
		"system": map[string]any{
			"fetch": (func(string, map[string]any) (map[string]any, error))(mockFetch),
		},
	})

	return &directSetupResult{client: client, calls: calls, live: false}
}

func resolveIdmap(um *sdk.UniversalManager, baseSDK *sdk.UniversalSDK, entityMap map[string]any) map[string]any {
	idEntries := []any{}
	for _, item := range vs.Items(entityMap) {
		entDef := core.ToMapAny(item[1])
		ename, _ := vs.GetProp(entDef, "name").(string)
		for i := 0; i < 3; i++ {
			idEntries = append(idEntries, fmt.Sprintf("%s%02d", ename, i))
		}
	}

	idmap := core.ToMapAny(vs.Transform(
		idEntries,
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	))
	if idmap == nil {
		idmap = map[string]any{}
	}

	// In live mode, discover real parent entity IDs.
	if os.Getenv("UNIVERSAL_TEST_LIVE") == "TRUE" {
		liveClient := sdk.NewUniversalSDK(um, map[string]any{
			"ref":   "voxgig-solardemo",
			"model": um.ResolveModel("voxgig-solardemo"),
		})

		for _, item := range vs.Items(entityMap) {
			eDef := core.ToMapAny(item[1])
			eName, _ := vs.GetProp(eDef, "name").(string)
			listOp := core.ToMapAny(vs.GetPath([]any{"op", "list"}, eDef))
			if listOp == nil {
				continue
			}
			targets, _ := vs.GetProp(listOp, "targets").([]any)
			if len(targets) == 0 {
				continue
			}
			listTarget := core.ToMapAny(targets[0])
			if listTarget == nil {
				continue
			}
			listParams, _ := vs.GetPath([]any{"args", "params"}, listTarget).([]any)
			if len(listParams) > 0 {
				continue // skip nested entities
			}
			parts, _ := vs.GetProp(listTarget, "parts").([]any)
			listPath := vs.Join(parts, "/", true)

			res, err := liveClient.Direct(map[string]any{
				"path": listPath, "method": "GET", "params": map[string]any{},
			})
			if err == nil && res["ok"] == true {
				if dataList, ok := res["data"].([]any); ok {
					for i := 0; i < len(dataList) && i < 3; i++ {
						if dm := core.ToMapAny(dataList[i]); dm != nil {
							ref := fmt.Sprintf("%s%02d", eName, i)
							idmap[ref] = dm["id"]
						}
					}
				}
			}
		}
	}

	return idmap
}
